fw/http.c

```diff
@@ -1320,6 +1478,112 @@ tfw_http_send_resp(TfwHttpReq *req, int status, const char *reason)
 		tfw_h1_send_resp(req, status);
 }
 
+static void
+tfw_http_send_resp2(TfwHttpReq *req, TfwStr *msg, int status)
+{
+	if (TFW_MSG_H2(req)) {
+		tfw_h2_send_resp2(req, msg, status, 0);
+	} else {
+		TfwCliConn *cli_conn = (TfwCliConn *)req->conn;
+
+		WARN_ONCE(!list_empty_careful(&req->msg.seq_list),
+			  "Request is already in seq_queue\n");
+		tfw_stream_unlink_msg(req->stream);
+		spin_lock(&cli_conn->seq_qlock);
+		list_add_tail(&req->msg.seq_list, &cli_conn->seq_queue);
+		spin_unlock(&cli_conn->seq_qlock);
+
+		tfw_h1_send_resp2(req, msg, status);
+	}
+}
+
+static void
+tfw_http_req_redir(TfwHttpReq *req, int status, const char *pattern)
+{
+	size_t url_len = strlen(pattern) + 1;
+	char *url = kmalloc(url_len, GFP_KERNEL);
```

Comment:
Ideally, if you can rewrite the code (having the comment at the below in mind) without the dynamic allocation, just using `TfwStr` fragments. However, this might be too complex and we also discussed that too many small fragments hurt performance, so, if you end up with the need of dynamic memory allocation, then just allocate a per-cpu static buffer, just like the current `g_buf` (maybe you'll have an idea how to aggregate them in an elegant way).